apps/system/js/system_banner.js

```diff
@@ -59,58 +72,88 @@
+
+      return this.getBanner().then(banner => {
+        // Clear the button and message containers.
+        var messageContainer = banner.querySelector('.messages');
+        var buttonContainer = banner.querySelector('.buttons');
+        messageContainer.innerHTML = '';
+        buttonContainer.innerHTML = '';
+
+        // Populate banner with messages.
+        if (!Array.isArray(messages)) {
+          messages = [messages];
+        }
+        messages.forEach((chunk) => {
+          var p = createChild(messageContainer, 'p');
+          setElementL10n(p, chunk);
         });
-      } else {
-        var span = document.createElement('span');
-        setElementL10n(span, message);
-        p.appendChild(span);
-      }
-
-      var button = banner.querySelector('button');
-      if (buttonParams) {
-        banner.dataset.button = true;
-        setElementL10n(button, buttonParams.label);
-        this._clickCallback = function() {
-          this._clicked = true;
-          buttonParams.callback();
-        }.bind(this);
-        button.addEventListener('click', this._clickCallback);
-      }
-
-      banner.addEventListener('animationend', function animationend() {
-        banner.removeEventListener('animationend', animationend);
-        banner.classList.remove('visible');
 
+        // Populate  buttons if specified.
         if (buttonParams) {
-          if (buttonParams.dismiss && !this._clicked) {
-            buttonParams.dismiss();
-          }
-          banner.dataset.button = false;
-          button.removeEventListener('click', this._clickCallback);
-          button.classList.remove('visible');
-          this.banner.parentNode.removeChild(this.banner);
-          this._banner = undefined;
+          var button = createChild(buttonContainer, 'span');
+          button.className = 'button';
+          setElementL10n(button, buttonParams.label);
+          button.addEventListener('click', () => {
+            this._clicked = true;
+            this.hide();
+            buttonParams.callback && buttonParams.callback();
+          });
         }
-      }.bind(this));
 
-      banner.classList.add('visible');
+        // Hide banner and invoke button dismiss() on click or timeout.
+        var hideBanner = () => {
+          this.hide();
+          if (!this._clicked) {
+            buttonParams && buttonParams.dismiss && buttonParams.dismiss();
+          }
+        };
+        banner.onclick = hideBanner;
+        setTimeout(hideBanner, SystemBanner.TIMEOUT);
+
+        document.getElementById('screen').appendChild(banner);
+        banner.show();
+      });
+    },
+
+    /**
+     * Hide the System Banner and remove it from the DOM.
+     */
+    hide: function() {
+      if (!this._showing) {
+        return;
+      }
+      this._showing = false;
+
+      return this.getBanner().then(banner => {
+        banner.hide();
+        banner.addEventListener('animationend', function removeBanner() {
```

Comment:
Shouldn't we add the listener before hiding? At the other hand, I think it would be nice if a 'hidden' event is triggered by the component instead of listening for animations here.